### ISSUE TITLE
Query Time Lookups for Dimensions in Druid

### ISSUE THREAD

**Issue by sid88in** [apache/incubator-superset#211]
This feature comes from http://druid.io/docs/latest/querying/lookups.html 

**Use Case:**
Have dimension -> values mapping at query time. In real world we can have small set of mappings approx 10 - 20 or  large mappings > 1000. When we visualize the data in Caravel, then the current version just shows the dimension values as integers if the data in druid is coming from fact table (metrics and dimensions). As we don't have joins in druid, we want query time lookups to get actual value of the dimension in charts and graphs. I have been able to hack the code and test this feature with map lookup feature. The query generator creates query in this form:
```
{
    "queryType": "groupBy",
    "dataSource": "XXX",
    "intervals": [
        "2016-03-16/2016-03-17"
    ],
    "granularity": "day",
    "dimensions": [
        {
            "type": "extraction",
            "dimension": "ABC",
            "outputName": "ABC",
            "extractionFn": {
                "type": "lookup",
                "lookup": {
                    "type": "map",
                    "map": {
                        "1": "sid",
                        "2": "gupta"
                    }
                },
                "retainMissingValue": true,
                "injective": false
            }
        }
    ],
    "aggregations": [
        {
            "type": "doubleSum",
            "fieldName": "XYZ",
            "name": "XYZ"
        },
        {
            "type": "longSum",
            "fieldName": "PQR",
            "name": "PQR"
        }
    ]
}
```
I want to discuss the design of this feature before coding it out and making it available for PR.
I am currently thinking of having 2 types of lookups: static lookup (small key value pairs) and dynamic lookups (jdbc lookups). For static lookup we can either have a function to preload the data from json mapping in the code or let users input key value mappings from the UI http://hostname:port/druiddatasourcemodelview/edit/2

Any thoughts? I still have to make jdbc lookups to work

**Comment by sashank-nistala** [comment]
Will this feature be implemented anytime soon?

**Comment by zhaown** [comment]
+1, Looking forward to this too.

**Comment by rohitkochar** [comment]
+1
We also need this feature

**Comment by kkalyan** [comment]
@sid88in can you share your work in progress code ?

**Comment by sid88in** [comment]
@kkalyan I moved from Yahoo to Glassdoor while working on Caravel :) Was able to hack the code to implement query time lookups in Caravel. Basically you will need to modify the query function (where the actually query gets formulated). I modified it to use map function and it worked.  I'll talk to Maxime to see what is his thought on this and we may want to engineer it properly to be used by everyone.

**Comment by mistercrunch** [comment]
I just added support for `dimensionsSpec` (only in master, not released yet) which exposes ways to do specify mapping on a per-column basis:
http://druid.io/docs/latest/querying/dimensionspecs.html

To use it, you have to create a new calculated column on the datasource and paste your valid `dimensionSpec` expression in the text box.

You can define your mapping as a hard coded JSON object in our dimensionSpec, and it also looks like you can define "namespace" mappings that are stored on all the nodes on the Druid cluster and you can reference them in `dimensionSpec`.
http://druid.io/docs/latest/querying/lookups.html

Maybe a complementary feature on the Caravel would be for a way to manage your Druid Lookups in Caravel and have Caravel sync them with the Druid cluster. This is not on our current roadmap but we get to it at some point.

**Comment by Igosuki** [comment]
@mistercrunch I used the feature today, and it seems the columns is ignored when building the result data from the Druid query (I can see the lookup dimension working if I do the query by myself to the Druid API).

**Comment by KostasChr** [comment]
I am happily using superset - and I have a calculated column on the datasource with a dimensionSpec expression that describes a query time lookup. 

Grouping works well but filtering does not on this calculated column...  Anyone succeeded in doing this? Maybe I need to update the query generator?

**Comment by ghulands** [comment]
This should be fixed via #4740
